Add the possibility to replay a service task after failure

erdemedeiros opened the pull request:
A new REST endpoint is available to replay a service task based on the related execution id and flow-node id.
```
POST /rb/admin/v1/executions/{executionId}/replay/service-task
{
   "flowNodeId": "flowNodeId"
}
```

Fixes https://github.com/Activiti/Activiti/issues/3971

erdemedeiros committed:
WIP: add test and first elements to replay a service task

igdianov committed:
fix: implement IntegrationRequestReplayer test

igdianov committed:
fix: add IntegrationErrorSender for CanFailConnector test

erdemedeiros committed:
fix copyright

erdemedeiros committed:
rename variables

erdemedeiros committed:
fix scope

erdemedeiros committed:
simplify replayer logic

erdemedeiros committed:
add REST endpoint for replay

igdianov committed:
fix: update MQServiceTaskBehaviorTest failing test

igdianov commented on activiti-cloud-runtime-bundle-service/activiti-cloud-services-runtime-bundle/activiti-cloud-services-rest-api/src/main/java/org/activiti/cloud/services/rest/api/RecoverController.java:
```diff
@@ -0,0 +1,32 @@
+
+@RequestMapping(value = "/admin/v1/recover",
+    produces = {MediaTypes.HAL_JSON_VALUE, MediaType.APPLICATION_JSON_VALUE })
+public interface RecoverController {
```
I think it would make sense to use better naming, i.e. `ServiceTaskAdminController` with base request endpoint `/admin/v1/service-task`. Then, use `POST /replay` path to map to replay method in the controller interface.

erdemedeiros replied:
Yes, it makes sense. I'll refactor it.

erdemedeiros replied:
After some discussion with the team, the new endpoint it's going to be:
```
POST /admin/v1/executions/{executionId}/replay/service-task
{
   "flowNodeId": "flowNodeId"
}
```
The parameter `executionId` is defining the resource, so it's going to be placed in the URL, while `flowNodeId` will go the the body, where we can eventually add extra information in the future (such as variables).

igdianov committed:
fix: add validation to ReplayServiceTaskRequest

igdianov committed:
fix: update ServicesAuditITConfiguration due to failing tests

igdianov committed:
fix: update ReplayServiceTaskRequest validation

igdianov commented on activiti-cloud-runtime-bundle-service/activiti-cloud-services-runtime-bundle/activiti-cloud-services-rest-api/src/main/java/org/activiti/cloud/services/rest/api/ReplayServiceTaskRequest.java:
```diff
@@ -0,0 +1,44 @@
+
+import javax.validation.constraints.NotEmpty;
+
+public class ReplayServiceTaskRequest {
```
I think this class should be moved to api module. We can also remove it in favor of using path variables to simplify , i.e. `POST /admin/v1/service-task/{executionId}/replay/{flowNodeId}`

erdemedeiros replied:
As long we don't include a specific method in the Java API it should be fine to keep it at the same level as the controller. Am I missing something?

igdianov replied:
I think it makes sense to keep it here.

erdemedeiros committed:
Rename rest endpoint

erdemedeiros committed:
improve test coverage

erdemedeiros committed:
Move flowNodeId to request body

It will make it easier to evolve and add extra information in the feature. For instance, set variables while replaying

erdemedeiros committed:
fix copyright

andrea-ligios reviewed (approved):
LGTM!
